internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -158,37 +264,84 @@ func (p *processRunner) build() (commandWrapper, error) {
 			return commandWrapper{}, err
 		}
 
-		cmd.Stderr = stderr
+		wrapper.stderr = stderr.Fd()
 
 		afterStartFuncs = append(afterStartFuncs, func() {
 			stderr.Close() //nolint:errcheck
 		})
 	} else {
-		cmd.Stderr = writer
+		wrapper.stderr = pw.Fd()
 	}
 
-	ctty, cttySet := p.opts.Ctty.Get()
-	if cttySet {
-		cmd.SysProcAttr = &syscall.SysProcAttr{
-			Setsid:  true,
-			Setctty: true,
-			Ctty:    ctty,
+	closeWriter = false
+
+	wrapper.launcher = launcher
+	wrapper.afterStart = func() {
+		for _, f := range afterStartFuncs {
+			f()
 		}
 	}
+	wrapper.afterTermination = closeLogging
+	wrapper.ctty = p.opts.Ctty
 
-	closeWriter = false
+	cgroupFdSupported := false
 
-	return commandWrapper{
-		cmd: cmd,
-		afterStart: func() {
-			for _, f := range afterStartFuncs {
-				f()
-			}
-		},
-		afterTermination: func() error {
-			return w.Close()
-		},
-	}, nil
+	platform, err := platform.CurrentPlatform()
+	if err == nil {
+		cgroupFdSupported = platform.Mode() != runtime.ModeContainer
+	}
+
+	cgroupPath := path.Join(constants.CgroupMountPath, cgroup.Path(p.opts.CgroupPath))
+
+	if cgroupFdSupported && p.opts.UID > 0 {
+		if err := os.Chown(cgroupPath, int(p.opts.UID), -1); err != nil {
+			cgroupFdSupported = false
+		}
+	}
+
+	// cgroupfd is more reliable, use it when possible
+	if cgroups.Mode() == cgroups.Unified && cgroupFdSupported {
+		cg, err := os.Open(cgroupPath)
+		if err == nil {
+			wrapper.cgroupFd = int(cg.Fd())
+		}
+	}
+
+	return wrapper, nil
```

Comment:
@smira since we return here do we need to tell the go runtime to keep the file open  :thinking:

Reply:
good question, let's pull in our expert @DmitriyMV

Reply:
From the docs for [os.File.Fd](https://pkg.go.dev/os#File.Fd):

>// If f is closed, the file descriptor becomes invalid.
// If f is garbage collected, a **finalizer may close** the file descriptor,
// **making it invalid**; see [runtime.SetFinalizer] for more information on when
// a finalizer might be run. On Unix systems this will cause the [File.SetDeadline]
// methods to stop working.

Emphasis mine. So yeah, GC is free to close this handle (and handles above) and code will break. I'm actually curious why you store `uintptr` instead of `*os.File` or `interface {io.Closer; Fd() uintptr}`? This would be much safer.

Reply:
okay, we can probably store them in commandWrapper. Is that going to be a sufficient lifetime? Perhaps so. Anyway, do we have to close those, or they're going to be cleaned up by kernel when child process exits? Going to take a look

Reply:
Just use `afterTermination` function field call to close them anyway.

Reply:
Will also store cgroupfd there as we must dispose of it after forking

Reply:
@DmitriyMV PTAL at the recent fix.

I found no issue with stdio, but replaced storing raw fds with *os.File, yes. Maybe I should add more non-nil checks to all the places they're used in, but likely not.

Added cgroupFile in the same way to close it

Reply:
Are you sure you need to close `stdin`/`stdout`/`stderr` right after forking ends and not after forked process completes?

Reply:
Well, it was done by the old code (the part invoking wrapperd) and tests are okay. I didn't reconsider this and am not sure how are such handed-off fd's handled after fork, but I believe it'd fail reliably in case our approach to it is wrong.